qa: parametrize logger usages

jdrueckert opened the pull request:
PMD Rule Docs: https://docs.pmd-code.org/pmd-doc-6.55.0/pmd_rules_java_bestpractices.html#guardlogstatement

Theoretically fixes rule findings by using substituting parameters (parametrized logs).
Practically, false-positive findings still show up due to PMD bug: https://github.com/pmd/pmd/issues/4703

jdrueckert committed:
qa: parametrize logger usages

jdrueckert commented:
Addresses the remaining findings that were not part of #5170

skaldarnar commented on engine/src/main/java/org/terasology/engine/config/RenderingDebugConfig.java:
```diff
@@ -108,6 +108,6 @@ public void setRenderSkeletons(boolean renderSkeletons) {
     @Override
     public void propertyChange(PropertyChangeEvent evt) {
-        logger.info("Set {} property to {}. ", evt.getPropertyName().toUpperCase(), evt.getNewValue()); // for debugging purposes
+        logger.info("Set {} property to {}.", evt.getPropertyName().toUpperCase(), evt.getNewValue()); // for debugging purposes
```
"for debugging purposes" - should this be a DEBUG log?

jdrueckert replied:
probably
```suggestion
        logger.debug("Set {} property to {}.", evt.getPropertyName().toUpperCase(), evt.getNewValue());
```

skaldarnar commented on engine/src/main/java/org/terasology/engine/core/TerasologyEngine.java:
```diff
@@ -266,7 +266,7 @@ private void verifyInitialisation() {
      */
     private void logEnvironmentInfo() {
-        logger.info(TerasologyVersion.getInstance().toString());
+        logger.info("{}", TerasologyVersion.getInstance().toString());
```
```suggestion
        logger.info("{}", TerasologyVersion.getInstance());
```

jdrueckert replied:
Interestingly, checkstyle didn't suggest to remove `toString()` here while in other places it did. I don't fully understand yet why it does it in some cases and doesn't in others. Any ideas?

skaldarnar replied:
Technically, you can `toString` on something that is already a `String` - maybe Checkstyle understands this and tells you that that is superflous? 🤔

jdrueckert replied:
Oh, that's a good point... so it tells me because `toString()` is not necessary on `Strings` but not because `toString()` is not necessary as it's implicitly called when logging 💡 makes sense 👍

skaldarnar commented on engine/src/main/java/org/terasology/engine/core/modes/StateLoading.java:
```diff
@@ -234,7 +234,7 @@ private void popStep() {
         if (!loadProcesses.isEmpty()) {
             current = loadProcesses.remove();
-            logger.debug(current.getMessage());
+            logger.debug("{}", current.getMessage());
```
This looks weird 🙄 Is this necessary because of some quirk in the rules?

The only this might do is defer the call to `toString` on whatever type `getMessage()` returns. If that is already string, we don't gain anything from this...

jdrueckert replied:
My understanding was that the parametrization is what avoids the call in case log level is not debug, so with
`logger.debug(current.getMessage())` the call to `getMessage()` is done even if log level is "info"
while with
`logger.debug("{}", current.getMessage())` the call to `getMessage()` is only done if log level is "debug" but not if it's "info"

Is my understanding incorrect?

Admittedly, in this particular case the call might not be particularly impactful.

skaldarnar replied:
> Is my understanding incorrect?

I think it indeed is. The only thing that is defered is the final call to `toString` I think.

Maybe this gets clearer if we write this part of the code like this:

```java
// executed regardless of the log-level
Message message = current.getMessage(); 
// message.toString() is only called if DEBUG logs are enabled (happens inside)
logger.debug("{}", message);
```

The parameterization can only help as far as Java itself allows it. Here, we are passing the value of `current.getMessage()` to the method, that is, we evaluate that part already when calling it.

To make it a truly lazy evaluation we'd need a supplier, either for the whole message or for the inidividual parameters (see https://github.com/MovingBlocks/Terasology/pull/5173#discussion_r1392577738). I don't think the following code works, but that's how it would look like:

```java
logger.debug(() -> current.getMessage());
```

Here, we pass a function that, if evaluated, will make the lookup on how to get the message from `current`.

skaldarnar commented on engine/src/main/java/org/terasology/engine/logic/characters/CharacterSystem.java:
```diff
@@ -315,35 +315,35 @@ private boolean isPredictionOfEventCorrect(EntityRef character, ActivationReques
             if (!event.getUsedOwnedEntity().exists()) {
-                String msg = "Denied activation attempt by {} since the used entity does not exist on the authority";
-                logger.info(msg, getPlayerNameFromCharacter(character));
+                logger.info("Denied activation attempt by {} since the used entity does not exist on the authority",
+                        getPlayerNameFromCharacter(character));
```
Does our logging framework allow for passing a `Supplier<String>` for the log message? Here, it would be nice to avoid the computation of `getPlayerNameFromCharacter(character)` until we actually have to log it. Maybe that's guarded sufficiently, though, and INFO logs are enabled by default currently 🤷

jdrueckert replied:
My understanding was that the parametrization does exactly that - avoid the computation until we actually have to log it. But I don't know for sure, so I could be misunderstanding how this works.

soloturn replied:
it is so @jdrueckert :
https://dennis-xlc.gitbooks.io/the-logback-manual/content/en/chapter-2-architecture/logger-appenders-and-layouts/parameterized-logging.html

jdrueckert replied:
@soloturn the parts mentioned in that manual I understood. It doesn't speak to method calls in the parameters, though.
But @skaldarnar's comment above explains it nicely 👍

soloturn replied:
ha, with you guys one learns something new all the time. there is indeed a fluent API to slf4j-2.0, @skaldarnar , which allows to pass suppliers. we are at slf4j-1.7 at the moment:
https://www.slf4j.org/manual.html#fluent

soloturn commented on engine/src/main/java/org/terasology/engine/rendering/dag/RenderTaskListGenerator.java:
```diff
@@ -183,7 +183,7 @@ public List<RenderPipelineTask> generateFrom(List<Node> orderedNodes) {
     private void logList(List<?> list) {
         for (Object object : list) {
-            logger.debug(object.toString());
+            logger.debug("{}", object.toString());
```
see above @skaldarnar "weird" comment.

jdrueckert replied:
```suggestion
            logger.debug("{}", object);
```

soloturn committed:
build: upgrade slf4j-2.0, permit passing suppliers to loggers (lazy evaluation) (#5174)

permit use of slf4j fluent api:
https://www.slf4j.org/manual.html#fluent

tried to upgrade to logback 1.3.x or logback 1.4.x, but code contains
code depending on ch.qos.logback.core.joran.spi.InterpretationContext:
engine/src/main/java/org/terasology/logback/RegexFilterAction.java

for engine tests though, an upgrade was necessary as the code uses
logback-classic directly.

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/dag/RenderTaskListGenerator.java:
```diff
@@ -48,15 +48,15 @@ private void logIntermediateRendererListForDebugging(List<Node> orderedNodes) {
                 // printing out individual desired state changes
                 for (StateChange desiredStateChange : node.getDesiredStateChanges()) {
-                    logger.info(desiredStateChange.toString());
+                    logger.info("{}", desiredStateChange.toString());
```
```suggestion
                    logger.info("{}", desiredStateChange);
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/nui/editor/layers/NUIEditorScreen.java:
```diff
@@ -279,7 +279,7 @@ public void selectAsset(ResourceUrn urn) {
                 content = new JsonParser().parse(reader).toString();
             } catch (IOException e) {
-                logger.error(String.format("Could not load asset source file for %s", urn.toString()), e);
+                logger.error("Could not load asset source file for {}", urn.toString(), e);
```
```suggestion
                logger.error("Could not load asset source file for {}", urn, e);
```

jdrueckert commented on engine/src/main/java/org/terasology/engine/rendering/nui/editor/layers/NUISkinEditorScreen.java:
```diff
@@ -315,7 +315,7 @@ public void selectAsset(ResourceUrn urn) {
                 content = new JsonParser().parse(reader).toString();
             } catch (IOException e) {
-                logger.error(String.format("Could not load asset source file for %s", urn.toString()), e);
+                logger.error("Could not load asset source file for {}", urn.toString(), e);
```
```suggestion
                logger.error("Could not load asset source file for {}", urn, e);
```

jdrueckert committed:
chore: remove superfluous toString() calls

Co-authored-by: Tobias Nett <skaldarnar@googlemail.com>

jdrueckert committed:
chore: remove superfluous toString calls

jdrueckert committed:
chore: reuse existing local variables in log

jdrueckert committed:
refactor: introduce new local variable to use in logic and logs

jdrueckert committed:
refactor: parametrize log

jdrueckert committed:
chore: fix minor typos in comment

jdrueckert committed:
chore: log full exception instead of only exception message

jdrueckert committed:
Merge branch 'develop' into qa/logger

jdrueckert committed:
Merge branch 'develop' into qa/logger

jdrueckert committed:
ci: retrigger

jdrueckert committed:
fix: variable scope

jdrueckert committed:
Merge branch 'develop' into qa/logger

skaldarnar commented on build-logic/src/main/kotlin/terasology-metrics.gradle.kts:
```diff
@@ -25,13 +25,13 @@ dependencies {
         because("allows use of EvaluatorFilter in logback.xml")
     }
-    testRuntimeOnly("org.slf4j:jul-to-slf4j:1.7.36") {
+    testRuntimeOnly("org.slf4j:jul-to-slf4j:2.0.9") {
```
Does that major version upgrade just work?

jdrueckert replied:
looked like it, yes
